### DISCUSSION TITLE
Support events (like mouse click) in Plotly charts

### DISCUSSION THREAD

**Discussion by YH20220622** [zauberzeug/nicegui#748] [Ideas / Feature Requests]
### Question

I can only find an example In the docs of Plotly python docs, which adds trace in go.FigureWidget() instead of go.Figure(). The doc says the click event will only emit in FigureWidget and JupyterLab env. 

Is there any way to customize click events in nicegui python?
Or some sample code using JavaScript to do this would also be helpful to me. 

I had some code for importing Plotly.js, then draw a chart in JavaScript using `ui.run_javascript`. It can work, but I don't think it is a good approach.
 
``` python
t_script = ''
with open('./srcs/plotly.min.js', 'r', encoding='utf-8') as f:
    t_script += f.read()
    await ui.run_javascript(f'{t_script}', timeout=1.0)
```

Thanks very much for any help!

**Comment by falkoschindler** [comment]
Interesting idea, @YH20220622!

Although you can register click handlers like in the following example, this is probably not what you're looking for:

```py
fig = go.Figure(go.Scatter(x=[1, 2, 3, 4], y=[1, 2, 3, 2.5]))
fig.update_layout(margin=dict(l=0, r=0, t=0, b=0))
p = ui.plotly(fig).on('click', lambda msg: ui.notify(msg), args=['clientX', 'clientY'])
```

This is just the plain JavaScript event, without relation to the Plotly figure. Capturing a real Plotly event is probably a bit more complicated. But maybe someone likes to try implementing a click handler in plotly.vue? Pull requests are always welcome.

**Reply by YH20220622** [reply]
Thanks for replying.
Definitely I would like a click event to get the picked point infomation and display something on the figure, so the JS event doesn't fit.

I post my code, which run the JavaScript to achieve picking event. I add header html to import plotly.js, and run plotly initializing script after button clicked. It works.
```python
from nicegui import ui, app

ui.add_head_html('<script src="https://cdn.plot.ly/plotly-2.20.0.min.js"></script>')

ui.add_body_html('<div id="myDiv" style="height: 500px" class="p-2 bg-blue-100 w-full"></div>')
ui.add_body_html('<h1 id="hoverinfo"></h1>')

js = """
        var myPlot = document.getElementById('myDiv'),
        hoverInfo = document.getElementById('hoverinfo'),
        N = 16,
        x = [0, 1, 2, 3, 4, 5, 6, 7],
        y1 = [7, 6, 5, 4, 3, 2, 1, 0],
        y2 = [10, 11, 12, 13, 14, 15, 16, 17],
        data = [{
            x: x,
            y: y1,
            type: 'scatter',
            name: 'Trial 1',
            mode: 'markers',
            marker: {
                size: 16
            }
        }, {
            x: x,
            y: y2,
            type: 'scatter',
            name: 'Trial 2',
            mode: 'markers',
            marker: {
                size: 16
            }
        }],
    layout = {
        hovermode: 'closest',
        title: 'Hover on Points'
    };

    Plotly.newPlot('myDiv', data, layout, {
        displaylogo: false
    });

    myPlot.on('plotly_hover', function(data) {
            var infotext = data.points.map(function(d) {
                return (d.data.name + ': x= ' + d.x + ', y= ' + d.y.toPrecision(3));
            });

            hoverInfo.innerHTML = infotext.join('<br/>');
        })
        .on('plotly_unhover', function(data) {
            hoverInfo.innerHTML = '';
        });
"""

async def on_click():
  await ui.run_javascript(js, respond=False)

ui.button('show_plotly', on_click=on_click)

ui.run(reload=False) 
```
After clicking the button , when I move mouse around the points, it will print its information on the label.

But I really don't like this way, it feels bad when I want to fetch data or use the fig deep inside.

**Comment by cramatt** [comment]
There are a few related discussions around this topic, see

- https://github.com/zauberzeug/nicegui/discussions/1818
- https://github.com/zauberzeug/nicegui/discussions/1598

using the above code sample its possible to kind of get this interaction but it seems overly complex, does not leverage the python api `ui.plotly()` and I'm not sure it allows for 2 way binding - aka on click in plotly, update state in python. 

One solution which could be nice and clean would be to have a method like `on_plotly()` which could be used in python like this: 

```py
def handle_hover(e):
  gui.hovered = e.value

ui.plotly(fig).on_plotly('plotly_hover', handler=lambda e: handle_hover(e))
```

If the api matched `on` with handler, debounce, etc that would be nice! 

I was considering attempting a PR for this, but not sure where to start. Is there an example of a 2 way binding element that is not just a quasar element?

Here are the relevant js and py files: 

- https://github.com/zauberzeug/nicegui/blob/main/nicegui/elements/plotly.vue
- https://github.com/zauberzeug/nicegui/blob/main/nicegui/elements/plotly.py

**Reply by cramatt** [reply]
Noticing that AG Grid component does what we are looking for (python handler for js interaction:) https://nicegui.io/documentation/ag_grid#respond_to_an_ag_grid_event

**Comment by falkoschindler** [comment]
I'll close this feature request as resolved in PR #2231.
